xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -89,49 +105,79 @@ public void start(final Listener2 listener) {
       listener.onError(Status.UNAVAILABLE.withDescription("Failed to bootstrap").withCause(e));
       return;
     }
-
-    List<ServerInfo> serverList = bootstrapInfo.getServers();
-    if (serverList.isEmpty()) {
+    if (bootstrapInfo.getServers().isEmpty()) {
       listener.onError(
           Status.UNAVAILABLE.withDescription("No traffic director provided by bootstrap"));
       return;
     }
 
-    // Currently we only support using the first server from bootstrap.
-    ServerInfo serverInfo = serverList.get(0);
+    final BootstrapInfo bootstrapInfoRef = bootstrapInfo;
+    XdsClientFactory xdsClientFactory = new XdsClientFactory() {
+      @Override
+      XdsClient createXdsClient() {
+        return
+            new XdsClientImpl(
+                bootstrapInfoRef.getServers(),
+                channelFactory,
+                bootstrapInfoRef.getNode(),
+                syncContext,
+                timeService,
+                backoffPolicyProvider,
+                stopwatchSupplier);
+      }
+    };
+    xdsClientPool = new RefCountedXdsClientObjectPool(xdsClientFactory);
+    xdsClient = xdsClientPool.getObject();
+    xdsClient.watchConfigData(hostName, port, new ConfigWatcher() {
+      @Override
+      public void onConfigChanged(ConfigUpdate update) {
+        String serviceConfig = "{\n"
+            + "  \"loadBalancingConfig\": [\n"
+            + "    {\n"
+            + "      \"cds_experimental\": {\n"
+            + "        \"cluster\": \"" + update.getClusterName() + "\"\n"
+            + "      }\n"
+            + "    }\n"
+            + "  ]\n"
+            + "}";
+        Map<String, ?> config;
+        try {
+          config = (Map<String, ?>) JsonParser.parse(serviceConfig);
+        } catch (IOException e) {
+          throw new AssertionError("Should never happen: invalid service config format");
```

Comment:
This can happen if `update.getClusterName()` contains illegal characters, so better call `onError()` and return than panic.

Reply:
Done.